Generics and other type hint improvements

scop opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Typed containers, TypedDict.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

scop committed:
Generic container type hint improvements

scop committed:
Use TypedDict to represent manifest

MartinHjelmare commented on homeassistant/loader.py:
```diff
@@ -54,7 +59,29 @@
+    """Integration manifest."""
+
+    name: str
+    disabled: Optional[str]
```
If `disabled` is present in the dict, will it always be a string?

scop replied:
I think so, it was already indirectly hinted as such before these changes:
```python
     @property
     def disabled(self) -> Optional[str]:
         """Return reason integration is disabled."""
         return cast(Optional[str], self.manifest.get("disabled"))
```

MartinHjelmare replied:
So then we shouldn't type it as `Optional`, right? We can add a comment that it's optional or maybe comment the other ones that are required.

scop replied:
Ah, you meant that. Guess we could go into looking for and making such changes here too, however my intention was to not make any changes of that nature, but just clean up and improve on missing parts. (Kind of an exception to that are `zeroconf` and `dhcp` which had incorrect `cast`s, but the casts went away entirely when moving to `TypedDict`.)

Thoughts?

MartinHjelmare replied:
We're not changing existing code. We're correctly typing the data in the TypedDict. The property `disabled` returns `Optional[str]` since the lookup of the manifest may return None if the key is missing. The `disabled`item in the manifest data though should not be typed as `Optional` since it should never be `None` if present.

scop replied:
It can be argued both ways, and neither is 100% correct. TypedDict cannot specify which specific attributes are required present, which not, and to preserve the "specification" which of them we want present, I chose to mark the non-required ones as Optional. But I don't feel that strongly about it so changed that field, ditto all others, as non-Optional, as this is not just about `disabled`.

Doing so required removing the legacy `"documentation": None`, grepping through the tree I found nothing that would break as a result.

scop committed:
Disable false pylint Popen unsubscriptable-object positive

scop committed:
Handle generic Popen on python 3.8

https://mypy.readthedocs.io/en/latest/runtime_troubles.html#using-classes-that-are-generic-in-stubs-but-not-at-runtime

scop committed:
Make all Manifest attributes non-Optional

MartinHjelmare reviewed (approved):
Nice!
